File: wire/src/mutation.rs
use serde::{Deserialize, Serialize};

pub const MANAGED_REQUEST_VERSION: u32 = 1;

/// Position of one applied mutation on the KV mutation topic: the fold
/// coordinate a barriered read waits for. A lease grant or renewal returns the
/// position at which the answering fold applied it, and a
/// [`crate::kv::KvGet`] carrying it as `min_position` must not be answered by
/// a fold that has not applied that position yet. Positions compare only
/// within one `topic_generation`: a generation mismatch is fail-closed
/// ([`crate::kv::KvError::Stale`]), never an implicit reset.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct MutationPosition {
    pub topic_generation: u64,
    pub partition: u32,
    pub offset: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ManagedRequestEnvelope {
    pub v: u32,
    pub operation_id: u128,
    #[serde(with = "crate::encoding::bin_bytes")]
    pub payload: Vec<u8>,
}

impl ManagedRequestEnvelope {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.v != MANAGED_REQUEST_VERSION {
            return Err("unsupported managed request version");
        }
        if self.operation_id == 0 {
            return Err("managed request operation id must not be zero");
        }
        Ok(())
    }
}

/// One managed command stored on a mutation topic. `payload` is the canonical
/// typed request encoded with the normal wire framing for `command_code`, so
/// append and fold share one request schema instead of maintaining a second
/// mutation representation.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct MutationCommandEnvelope {
    pub v: u32,
    pub operation_id: u128,
    pub timestamp_micros: u64,
    pub command_code: u32,
    #[serde(with = "crate::encoding::bin_bytes")]
    pub payload: Vec<u8>,
}

impl MutationCommandEnvelope {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.v == 0 {
            return Err("mutation command version must not be zero");
        }
        if self.operation_id == 0 {
            return Err("mutation command operation id must not be zero");
        }
        Ok(())
    }
}

#[cfg(all(test, feature = "cbor"))]
mod tests {
    use super::*;
    use crate::codes::{AGDX_KV_SET_CODE, KV_OP_VERSION};
    use crate::framing::{decode_named, encode_named};

    #[test]
    fn given_a_command_envelope_when_round_tripped_then_should_preserve_request_bytes() {
        let envelope = MutationCommandEnvelope {
            v: KV_OP_VERSION,
            operation_id: 42,
            timestamp_micros: 1_700_000_000_000_000,
            command_code: AGDX_KV_SET_CODE,
            payload: vec![0, 1, 2, 255],
        };
        let bytes = encode_named(&envelope).expect("encodes");
        let back: MutationCommandEnvelope = decode_named(&bytes).expect("decodes");
        back.validate().expect("the envelope is valid");
        assert_eq!(back.v, KV_OP_VERSION);
        assert_eq!(back.operation_id, 42);
        assert_eq!(back.timestamp_micros, 1_700_000_000_000_000);
        assert_eq!(back.command_code, AGDX_KV_SET_CODE);
        assert_eq!(back.payload, vec![0, 1, 2, 255]);
    }

    #[test]
    fn given_a_managed_request_when_round_tripped_then_should_preserve_operation_identity() {
        let request = ManagedRequestEnvelope {
            v: MANAGED_REQUEST_VERSION,
            operation_id: u128::MAX,
            payload: vec![4, 3, 2, 1],
        };

        let bytes = encode_named(&request).expect("managed request encodes");
        let back: ManagedRequestEnvelope = decode_named(&bytes).expect("managed request decodes");

        assert_eq!(back, request);
        back.validate().expect("the request is valid");
    }

    #[test]
    fn given_a_zero_operation_identity_when_validated_then_should_reject_it() {
        let request = ManagedRequestEnvelope {
            v: MANAGED_REQUEST_VERSION,
            operation_id: 0,
            payload: Vec::new(),
        };
        let command = MutationCommandEnvelope {
            v: KV_OP_VERSION,
            operation_id: 0,
            timestamp_micros: 1,
            command_code: AGDX_KV_SET_CODE,
            payload: Vec::new(),
        };

        assert!(request.validate().is_err());
        assert!(command.validate().is_err());
    }
}

File: foreign/typescript/src/wire/mutation.ts
import { CodecError } from "../client/errors.js"
import { type CborMap, field } from "./cbor.js"

export const MANAGED_REQUEST_VERSION = 1

export interface ManagedRequestEnvelope {
  readonly v: number
  readonly operationId: bigint
  readonly payload: Uint8Array
}

export function encodeManagedRequestEnvelope(
  envelope: ManagedRequestEnvelope
): Map<string, unknown> {
  return new Map<string, unknown>([
    ["v", BigInt(envelope.v)],
    ["operation_id", envelope.operationId],
    ["payload", envelope.payload]
  ])
}

export function decodeManagedRequestEnvelope(
  map: CborMap,
  context: string
): ManagedRequestEnvelope {
  const version = field.requiredU32(map, "v", context)
  if (version !== MANAGED_REQUEST_VERSION) {
    throw new CodecError("unsupported managed request version", "managed", "v")
  }
  const operationId = field.requiredU128(map, "operation_id", context)
  if (operationId === 0n) {
    throw new CodecError("managed request operation id must not be zero", "managed", "operation_id")
  }
  return {
    v: version,
    operationId,
    payload: field.requiredBytes(map, "payload", context)
  }
}

/** Position of one applied mutation on the KV mutation topic: the fold
 * coordinate a barriered read waits for. Positions compare only within one
 * `topicGeneration`. A generation mismatch is fail-closed. */
export interface MutationPosition {
  readonly topicGeneration: bigint
  readonly partition: number
  readonly offset: bigint
}

export function encodeMutationPosition(position: MutationPosition): Map<string, unknown> {
  return new Map<string, unknown>([
    ["topic_generation", position.topicGeneration],
    ["partition", BigInt(position.partition)],
    ["offset", position.offset]
  ])
}

export function decodeMutationPosition(map: CborMap, context: string): MutationPosition {
  return {
    topicGeneration: field.requiredU64(map, "topic_generation", context),
    partition: field.requiredU32(map, "partition", context),
    offset: field.requiredU64(map, "offset", context)
  }
}

export interface MutationCommandEnvelope {
  readonly v: number
  readonly operationId: bigint
  readonly timestampMicros: bigint
  readonly commandCode: number
  readonly payload: Uint8Array
}

export function encodeMutationCommandEnvelope(
  envelope: MutationCommandEnvelope
): Map<string, unknown> {
  return new Map<string, unknown>([
    ["v", BigInt(envelope.v)],
    ["operation_id", envelope.operationId],
    ["timestamp_micros", envelope.timestampMicros],
    ["command_code", BigInt(envelope.commandCode)],
    ["payload", envelope.payload]
  ])
}

export function decodeMutationCommandEnvelope(
  map: CborMap,
  context: string
): MutationCommandEnvelope {
  const version = field.requiredU32(map, "v", context)
  if (version === 0) {
    throw new CodecError("mutation command version must not be zero", "mutation", "v")
  }
  const operationId = field.requiredU128(map, "operation_id", context)
  if (operationId === 0n) {
    throw new CodecError(
      "mutation command operation id must not be zero",
      "mutation",
      "operation_id"
    )
  }
  return {
    v: version,
    operationId,
    timestampMicros: field.requiredU64(map, "timestamp_micros", context),
    commandCode: field.requiredU32(map, "command_code", context),
    payload: field.requiredBytes(map, "payload", context)
  }
}
